src/modm/processing/fiber/mutex.hpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2024, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include "fiber.hpp"
+#include <modm/architecture/interface/atomic_lock.hpp>
+#include <atomic>
+#include <mutex>
+
+namespace modm::fiber
+{
+
+/// @ingroup modm_processing_fiber
+/// @{
+
+/// Implements the `std::mutex` interface for fibers.
+/// @see https://en.cppreference.com/w/cpp/thread/mutex
+class mutex
+{
+	mutex(const mutex&) = delete;
+	mutex& operator=(const mutex&) = delete;
+
+	std::atomic_bool locked{false};
+public:
+	constexpr mutex() = default;
+
+	[[nodiscard]]
+	bool inline
+	try_lock()
+	{
+		bool expected{false};
+		return locked.compare_exchange_strong(expected, true);
```

Comment:
```suggestion
		return locked.compare_exchange_strong(expected, true, std::memory_order_acquire, std::memory_order_relaxed);
```

A mutex only needs acquire-release ordering, acquire on lock and release on unlock. Acquire means no access can be reordered before locking the mutex and release says no access can be moved behind the unlock.
In case the compare-and-swap operation fails, no other operations are synchronized, so relaxed is fine.

Reply:
Ok, still not entirely sure my choices were correct, but I've now added a memory order to all operations. I'm mostly unsure about my choices in condition variable.

Reply:
I'll try to double check it, most likely on the weekend. I'll be quite busy this week.